### PR TITLE
Turtlebot3 demo

### PR DESCRIPTION
**Pull request by maximilianwulf**
- This demo provides a simulation environment for elevation mapping
 - To start it, use roslaunch elevation_mapping_demos turtlesim3_waffle_demo.launch
 - Elevation mapping will subscribe to the simulated point cloud
 - The robot can be moved around
 - Since the resolution of the realsense is a bit high in the original, the urdf has been copied such that the parameters can be adapted
 - tf_to_pose_publisher.py is a copy of ground_truth_pose_publisher.py but with rosparams

elevation_mapping_turtlebot3_compiled_in_release

### CODE REVIEW

**Comment by maximilianwulf** [comment]
@goodfaiter could you quickly test the demo based on the descriptions in the README?

File: elevation_mapping_demos/rviz/turtlebot3_waffle_demo.rviz [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+    collapsed: false
+  Time:
+    collapsed: false
+  Tool Properties:
+    collapsed: false
+  Views:
+    collapsed: true
+  Width: 2493
+  X: 1987
+  Y: 27
```
Comment (remod) [inline elevation_mapping_demos/rviz/turtlebot3_waffle_demo.rviz]: I would set Height, Width, X, and Y manually (X looks like you run it on your external display ;) )

Comment (maximilianwulf) [reply elevation_mapping_demos/rviz/turtlebot3_waffle_demo.rviz]: Thank you for the comment. I ran it on my external display :)

File: README.md [hunk 2]
```diff
@@ -90,6 +90,22 @@ Run the unit tests with
 
 In order to get the Robot-Centric Elevation Mapping to run with your robot, you will need to adapt a few parameters. It is the easiest if duplicate and adapt all the parameter files that you need to change from the `elevation_mapping_demos` package (e.g. the `simple_demo` example). These are specifically the parameter files in `config` and the launch file from the `launch` folder.
 
+### Turtlebot3 waffle simulation
+
+A full running pipeline is available together with the turtlebot3 simulation. It can be used to test the elevation mapping on a computer without access to a real robot platform or as a starting point for further integration. The turtlebot3 simulation needs to be downloaded first and can be easily done with
```
Comment (YoshuaNava) [inline README.md]: I would rephrase this as
```
A running example is provided, making use of the Turtlebot3 simulation environment. This example can be used to test elevation mapping, or as a starting point for further integration. 

To start with, the Turtlebot3 simulation dependencies need to be installed:
    sudo apt install ros-melodic-turtlebot3*

The elevation mapping demo together with the turtlebot3 simulation can be started with
    roslaunch elevation_mapping_demos turtlesim3_waffle_demo.launch

To control the robot with a keyboard, a new terminal window needs to be opened (remember to source your ROS environment). Then run

    export TURTLEBOT3_MODEL=waffle
    roslaunch turtlebot3_teleop turtlebot3_teleop_key.launch

Velocity inputs can be sent to the robot by pressing the keys `a`, `w`,`d`, `x`. To stop the robot completely, press `s`.
```

File: elevation_mapping_demos/scripts/tf_to_pose_publisher.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
+#!/usr/bin/env python
+import rospy
+import geometry_msgs.msg
+import std_msgs
+import tf
+
+def callback(newPose):
+    global publisher, tfListener, from_frame, to_frame
```
Comment (YoshuaNava) [inline elevation_mapping_demos/scripts/tf_to_pose_publisher.py]: Docstring for callback? [[Source](https://realpython.com/python-pep8/)]

File: README.md (hunk 2 above)
Comment (maximilianwulf) [reply README.md]: Thank you for the suggestions, I added them.

File: elevation_mapping_demos/scripts/tf_to_pose_publisher.py (hunk 3 above)
Comment (maximilianwulf) [reply elevation_mapping_demos/scripts/tf_to_pose_publisher.py]: I added docstring to the function and more comments to the code.

File: README.md [hunk 4]
```diff
@@ -90,6 +90,24 @@ Run the unit tests with
 
 In order to get the Robot-Centric Elevation Mapping to run with your robot, you will need to adapt a few parameters. It is the easiest if duplicate and adapt all the parameter files that you need to change from the `elevation_mapping_demos` package (e.g. the `simple_demo` example). These are specifically the parameter files in `config` and the launch file from the `launch` folder.
 
+### Turtlebot3 waffle simulation
```
Comment (pfankhauser) [inline README.md]: TurtleBot3 Waffle Simulation

File: elevation_mapping_demos/scripts/tf_to_pose_publisher.py (hunk 3 above)
Comment (YoshuaNava) [reply elevation_mapping_demos/scripts/tf_to_pose_publisher.py]: Solid :+1:

**Comment by maximilianwulf** [comment]
@pfankhauser, thank you for your review. 
> Is there a way that we don't need to duplicate the Python script?

The original script listens to the tf between the 'optitrak' and 'total_station' frames and publishes it as a pose. The new one takes the frame_ids as rosparams. 
In my opinion, the best solution would be to replace it with the new one. The only thing I don't know is if some code relies on it. Do you know? 
It was added in April 2015 and I guess that it was used in conjunction with ground_truth_demo.launch. But this launch does not call the script.

>  In the GIF, holes appear at various positions, is there an explanation for this?

I don't know yet, I will look into it. 

> Make sure to add yourself as co-author and maintainer =)

Thank you very much :)

**Comment by pfankhauser** [comment]
> The original script listens to the tf between the 'optitrak' and 'total_station' frames and publishes it as a pose. The new one takes the frame_ids as rosparams.
> In my opinion, the best solution would be to replace it with the new one. The only thing I don't know is if some code relies on it. Do you know?
> It was added in April 2015 and I guess that it was used in conjunction with ground_truth_demo.launch. But this launch does not call the script.

I wrote this for generating figures for a paper, but it's not used in the active code. Please replace the old one with the new version with rosparam. thanks!

File: README.md [hunk 5]
```diff
@@ -7,8 +7,9 @@ This is a [ROS] package developed for elevation mapping with a mobile robot. The
 The Robot-Centric Elevation Mapping packages have been tested under ROS Melodic and Ubuntu 18.04. This is research code, expect that it changes often and any fitness for a particular purpose is disclaimed.
 
 **Author: Péter Fankhauser<br />
+Co-Author: Maximilian Wulf<br />
 Affiliation: [ANYbotics](https://www.anybotics.com/)<br />
-Maintainer: Péter Fankhauser, pfankhauser@anybotics.com<br />**
+Maintainer: Maximilian Wulf, mwulf@anybotics.com<br />**
```
Comment (YoshuaNava) [inline README.md]: I think you also have to add your data to the package.xml.
